### PR TITLE
feat(grid): #RDV-23 creation grid (front)

### PR DESCRIPTION
**Pull request by niko-amosse**
## Preview

image

image

image

## Describe your changes

This PR introduces the initial setup for the frontend, including a 3-tab interface. The setup includes the following key components and functionality:

- **Tabs Interface**: Three main tabs are created as the core navigation elements.
  - One of these tabs contains a **button** that opens a **modal**.

- **Modal**: The modal consists of a two-page layout designed for creating a grid.
  - The modal includes:
    - **Containers**: Organized structures for managing modal content and layout.
    - **Components**: Reusable and modular components specifically built for the modal’s functionality.
    - **Providers**: Context and state providers necessary for managing the modal’s data flow and state across its two-page structure.
    - **Error Handling and Grid Creation Testing**:
  - Implemented and tested error handling logic for form input validation.
  - Successfully tested the grid creation functionality within the modal.

This setup lays the groundwork for further functionality and ensures modular, organized code for managing the modal and other UI elements in the frontend.

## Issue ticket number and link

[RDV-23](https://jira.support-ent.fr/browse/RDV-23)

## Checklist before requesting a review (magic string, indentation, comment/documentation...)
- [ ] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: frontend/src/containers/GridModal/index.tsx [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+export const GridModal: FC<GridModalProps> = ({
+  isOpen,
+  handleClose,
+  gridModalType,
+}) => {
+  const { t } = useTranslation("appointments");
+  const [page, setPage] = useState<PAGE_TYPE>(PAGE_TYPE.FIRST);
+
+  const isDisplayFirstPage =
```
Comment (ThomasSevagen) [inline frontend/src/containers/GridModal/index.tsx:23]: t'as vraiment besoin de ces deux const? tu peux pas en faire juste une seule et faire une ternaire avec?
CF com l 43

File: frontend/src/containers/GridModal/index.tsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+              {t("appointments.create.grid.title")}
+            </Typography>
+            <IconButton onClick={handleClose}>
+              <CloseIcon />
+            </IconButton>
+          </Box>
+          <Typography sx={{ fontStyle: "italic" }}>
+            {t("appointments.grid.required")}
+          </Typography>
+          {isDisplayFirstPage && <FirstPageGridModal />}
```
Comment (ThomasSevagen) [inline frontend/src/containers/GridModal/index.tsx:43]: `{isDisplayFirstPage ?<FirstPageGridModal />:<SecondPageGridModal />}`

File: frontend/src/core/enums.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+export enum SLOT_DURATION {
+    FIVETEEN_MINUTES = "15 MIN",
+    THIRTY_MINUTES = "30 MIN",
+    FOURTYFIVE_MINUTES = "45 MIN",
+    ONE_HOUR = "1 H",
+}
+
+export enum DAY {
```
Comment (ThomasSevagen) [inline frontend/src/core/enums.ts:8]: `export enum DAY {
    MONDAY = "MONDAY",
etc
}`
faire une fonction get...ByDays: swicth case qui transforme tes enums en keysI18N

File: frontend/src/core/types.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    minute: number | null;
+}
+
+export interface Slot {
+
+    begin: Time;
+    end: Time;
+}
+
+export interface WeekSlotsModel {
```
Comment (ThomasSevagen) [inline frontend/src/core/types.ts:14]: imaginons que ce soit pas des jours de semaine et que tu aies 250 keys

`type WorkingDay = Exclude<DAY, DAY.SUNDAY>;`

```
interface WeekSlotsModel {
  [K in WorkingDay]: Slot[];
}
```

File: frontend/src/providers/GlobalProvider/types.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+import { MODAL_TYPE } from "./enum";
+
+export interface GlobalProviderContextProps {
+  displayModals: DisplayModalsState;
+  setDisplayModals: Dispatch<SetStateAction<DisplayModalsState>>;
+  handleDisplayModal: (modalType: MODAL_TYPE) => void;
+}
+
+export interface GlobalProviderProps {
+  children: React.ReactNode;
```
Comment (ThomasSevagen) [inline frontend/src/providers/GlobalProvider/types.ts:12]: ReactNode direct

File: frontend/src/providers/GridModalProvider/index.tsx [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  
+  const mockPublicList: Public[] = [
+    { name: "Public 1", id: "1" },
+    { name: "Public 2", id: "2" },
+    { name: "Public 3", id: "3" },
+    { name: "Public 4", id: "4" },
+    { name: "Public 5", id: "5" },
+  ];
+
+  const slotDurationOptions : SLOT_DURATION[] = [
```
Comment (ThomasSevagen) [inline frontend/src/providers/GridModalProvider/index.tsx:56]: dans un utils

File: frontend/src/providers/GridModalProvider/index.tsx [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    SLOT_DURATION.FIVETEEN_MINUTES,
+    SLOT_DURATION.THIRTY_MINUTES,
+    SLOT_DURATION.FOURTYFIVE_MINUTES,
+    SLOT_DURATION.ONE_HOUR,
+  ];
+    
+  const periodicityOptions: PERIODICITY[] = [
```
Comment (ThomasSevagen) [inline frontend/src/providers/GridModalProvider/index.tsx:63]: pareil

File: frontend/src/routes/home/index.tsx [hunk 8]
```diff
@@ -1,6 +1,13 @@
 import React from "react";
 
+import { Box, Tab, Tabs, Typography } from "@mui/material";
 import { ID } from "edifice-ts-client";
+import { useTranslation } from "react-i18next";
+
+import { contentStyle, homeStyle, tabsStyle, titleStyle } from "./style";
+import { FindAppointments } from "~/containers/FindAppointments";
+import { MyAppointments } from "~/containers/MyAppointments";
+import { MyAvailability } from "~/containers/MyAvailability";
```
Comment (ThomasSevagen) [inline frontend/src/routes/home/index.tsx:11]: on export les types.
on enleve tous les React. et on importe directement les méthodes

File: frontend/src/components/DailySlot/index.tsx [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    errorInputs: { slots },
+    updateGridModalInputs: { handleDeleteSlot, handleSlotChange },
+  } = useGridModalProvider();
+
+  const isSlotError =
+    slots.ids.some((item) => item === slot.id) && (!slot.begin || !slot.end);
+
+  return (
+    <Box
+      sx={{
```
Comment (ThomasSevagen) [inline frontend/src/components/DailySlot/index.tsx:42]: styled component ici on fait pas de conditionnel dans le sx (tu demandera à flo ou alice ils t'expliqueront)

File: frontend/src/components/DailySlot/index.tsx [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+              {formatTime(time)}
+            </MenuItem>
+          ))}
+        </Select>
+      </FormControl>
+      <Typography>{t("appointments.to") + " :"}</Typography>
+      <FormControl variant="filled" sx={formControlStyle}>
+        <Select
+          onChange={(e) =>
+            handleSlotChange(day, slot.id, e.target.value as string, "end")
```
Comment (ThomasSevagen) [inline frontend/src/components/DailySlot/index.tsx:73]: as string obligatoire?

File: frontend/src/components/WeekSlots/index.tsx [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    const errors: Record<DAY, boolean> = {} as Record<DAY, boolean>;
+    for (const day of Object.keys(inputs.weekSlots) as DAY[]) {
+      errors[day] = inputs.weekSlots[day].some(
+        (slot) => slots.ids.includes(slot.id) && (!slot.begin || !slot.end),
+      );
+    }
+    return errors;
+  }, [inputs.weekSlots, slots]);
+
+  console.log(slots.ids);
```
Comment (ThomasSevagen) [inline frontend/src/components/WeekSlots/index.tsx:44]: log

File: frontend/src/containers/FirstPageGridModal/index.tsx [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+      <Box sx={spaceBetweenBoxStyle}>
+        <TextField
+          id="grid-name"
+          label={
+            t("appointments.grid.name") +
+            " * " +
+            t("appointments.grid.visible.all")
+          }
+          variant="outlined"
+          sx={{ width: "100%" }}
```
Comment (ThomasSevagen) [inline frontend/src/containers/FirstPageGridModal/index.tsx:54]: regarde si dans ce cpt il y a pas la prop `fullWidth` desfois elle y est

File: frontend/src/containers/FirstPageGridModal/index.tsx [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+          onBlur={handleNameBlur}
+          error={!!errorInputs.name}
+          helperText={t(errorInputs.name)}
+        />
+        <Box sx={flexEndBoxStyle}>
+          <Typography>{t("appointments.grid.color") + " * "}</Typography>
+          <ColorPicker />
+        </Box>
+      </Box>
+      <FormControl sx={{ width: "100%" }}>
```
Comment (ThomasSevagen) [inline frontend/src/containers/FirstPageGridModal/index.tsx:66]: same

File: frontend/src/containers/GridModal/index.tsx [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+        error: "",
+      },
+    };
+    setErrorInputs(newErrors);
+    if (newErrors.name || newErrors.visioLink) return;
+    setPage(PAGE_TYPE.SECOND);
+  };
+
+  const handleSubmit = async () => {
+    const newErrors = {
```
Comment (ThomasSevagen) [inline frontend/src/containers/GridModal/index.tsx:72]: parfait

File: frontend/src/containers/GridModal/index.tsx [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+      newErrors.visioLink ||
+      newErrors.validityPeriod ||
+      newErrors.weekSlots ||
+      newErrors.slots.ids.length
+    )
+      return;
+
+    const payload: GridPayload = gridInputsToGridPayload(inputs);
+    const response = await createGrid(payload);
+    console.log(response);
```
Comment (ThomasSevagen) [inline frontend/src/containers/GridModal/index.tsx:91]: log

File: frontend/src/containers/MyAvailability/index.tsx [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+import { spaceBetweenBoxStyle } from "~/styles/boxStyles";
+
+export const MyAvailability: FC = () => {
+  const { t } = useTranslation("appointments");
+  const {
+    displayModals: { grid },
+    handleDisplayModal,
+  } = useGlobalProvider();
+
+  const [gridModalType] = useState<GRID_MODAL_TYPE>(GRID_MODAL_TYPE.CREATION);
```
Comment (ThomasSevagen) [inline frontend/src/containers/MyAvailability/index.tsx:21]: à quoi ça sert de créer un state si t'as jamais besoin de le set?

File: frontend/src/containers/SecondPageGridModal/index.tsx [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+        </Box>
+        <RangeDatePicker />
+      </Box>
+      <Box sx={itemStyle}>
+        <Typography>{t("appointments.grid.slot.duration") + " *"}</Typography>
+        <Box sx={validityPeriodStyle}>
+          <ToggleButtonGroup exclusive value={inputs.slotDuration}>
+            {slotDurationOptions.map((option) => (
+              <ToggleButton
+                key={option}
```
Comment (ThomasSevagen) [inline frontend/src/containers/SecondPageGridModal/index.tsx:59]: on est sur de la singularité de option?

File: frontend/src/containers/SecondPageGridModal/index.tsx [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+          </ToggleButtonGroup>
+        </Box>
+      </Box>
+      <Box sx={itemStyle}>
+        <Typography>{t("appointments.grid.periodicity") + " *"}</Typography>
+        <Box sx={validityPeriodStyle}>
+          <ToggleButtonGroup exclusive value={inputs.periodicity}>
+            {periodicityOptions.map((option) => (
+              <ToggleButton
+                key={option}
```
Comment (ThomasSevagen) [inline frontend/src/containers/SecondPageGridModal/index.tsx:76]: same

File: frontend/src/hooks/useUpdateGridInputs.ts [hunk 19]
```diff
@@ -0,0 +1,222 @@
...
+  const handleNameChange = (e: ChangeEvent<HTMLInputElement>) => {
+    updateInputField("name", e.target.value);
+    updateErrorInputs("name", "");
+  };
+
+  const handleColorChange = (color: HexaColor) => {
+    updateInputField("color", color);
+  };
+
+  const handleStructureChange = (event: SelectChangeEvent) => {
```
Comment (ThomasSevagen) [inline frontend/src/hooks/useUpdateGridInputs.ts:66]: vu la redondance de certaines fonctions dans ce hook je pense qu'il y a moyen de refacto

File: frontend/src/providers/GlobalProvider/utils.ts [hunk 20]
```diff
@@ -0,0 +1,7 @@
+import { MODAL_TYPE } from "./enum";
+import { DisplayModalsState } from "./types";
+
+export const initialDisplayModalsState: DisplayModalsState = {
```
Comment (ThomasSevagen) [inline frontend/src/providers/GlobalProvider/utils.ts:4]: <3

File: frontend/src/components/DailySlot/index.tsx (hunk 9 above)
Comment (niko-amosse) [reply frontend/src/components/DailySlot/index.tsx:42]: fait, en effet c'est styled

File: frontend/src/components/DailySlot/index.tsx (hunk 10 above)
Comment (niko-amosse) [reply frontend/src/components/DailySlot/index.tsx:73]: non c'est viré

File: frontend/src/components/DailySlot/utils.ts [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+  return time.isAfter(start) && time.isBefore(end);
+};
+
+export const generateTimeSlots = (
+  intervalInMinutes: number,
+  isEnd: boolean,
+): TimeObject[] => {
+  const start = dayjs().startOf("day");
+
+  const totalMinutesInDay = 24 * 60;
```
[the remark this answers is not in the document]
Comment (niko-amosse) [reply frontend/src/components/DailySlot/utils.ts:36]: fait

File: frontend/src/components/WeekSlots/index.tsx (hunk 11 above)
Comment (niko-amosse) [reply frontend/src/components/WeekSlots/index.tsx:44]: viré

File: frontend/src/containers/FirstPageGridModal/index.tsx (hunk 12 above)
Comment (niko-amosse) [reply frontend/src/containers/FirstPageGridModal/index.tsx:54]: rempacé

File: frontend/src/containers/GridModal/index.tsx (hunk 15 above)
Comment (niko-amosse) [reply frontend/src/containers/GridModal/index.tsx:91]: viré

File: frontend/src/containers/SecondPageGridModal/index.tsx (hunk 17 above)
Comment (niko-amosse) [reply frontend/src/containers/SecondPageGridModal/index.tsx:59]: yep

File: frontend/src/containers/SecondPageGridModal/index.tsx (hunk 18 above)
Comment (niko-amosse) [reply frontend/src/containers/SecondPageGridModal/index.tsx:76]: idem c'est les valeurs des enums donc unique

File: JenkinsFile [hunk 22]
```diff
@@ -27,5 +27,18 @@ pipeline {
         }
       }
     }
+
+    stage('Stop Containers') {
```
[the remark this answers is not in the document]
Comment (niko-amosse) [reply JenkinsFile:31]: viré

**Pull request merged**